### PR TITLE
MM-25040: Omits some groups from channels where the team is group-con…

### PR DESCRIPTION
**Pull request by mkraft**
…strained.

### CODE REVIEW

**Comment by gsagula** [comment]
This PR is out of my domain of expertise. Let me know if you think that my review here is still valid or relevant.

**Comment by mkraft** [comment]
> This PR is out of my domain of expertise. Let me know if you think that my review here is still valid or relevant.

I'll re-assign it.

**Comment by mkraft** [comment]
> Code LGTM. I would just say that we write 2 more test that don't use only the length but also check the actual groups we get back. Ensure that the groups we are getting back are what we are expecting. Once for a team that is not group constrained and once for a team that is group constrained. Thoughts?

Yes I think you're right. I'll add them.

File: src/selectors/entities/groups.test.js [hunk 1]
```diff
@@ -134,7 +137,12 @@ describe('Selectors.Groups', () => {
 
     it('getGroupsNotAssociatedToChannel', () => {
         const expected = Object.entries(testState.entities.groups.groups).filter(([groupID]) => !channelAssociatedGroupIDs.includes(groupID)).map(([, group]) => group);
-        assert.deepEqual(Selectors.getGroupsNotAssociatedToChannel(testState, channelID), expected);
+        assert.deepEqual(Selectors.getGroupsNotAssociatedToChannel(testState, channelID, teamID), expected);
+
+        const cloneState = JSON.parse(JSON.stringify(testState));
+        cloneState.entities.teams.teams[teamID].group_constrained = true;
+        cloneState.entities.teams.groupsAssociatedToTeam[teamID].ids = [expectedAssociatedGroupID1];
```
Comment (fmunshi) [inline src/selectors/entities/groups.test.js:144]: can we just use destructuring to clone the new state?

Comment (mkraft) [reply src/selectors/entities/groups.test.js:144]: > can we just use destructuring to clone the new state?

I tried that first but our `deepFreezeAndThrowOnMutation` function rejected it. Let me try again because it *should* work.

Comment (mkraft) [reply src/selectors/entities/groups.test.js:144]: Looking into it a bit more it looks like the custom `__defineSetter__` methods are cloned across when doing `const cloneState = {...testState}`, which explains why it doesn't work. To clone the object like that I think we would have to re-define the methods, something like this:

```js
cloneState.entities.teams.teams[teamID].__defineSetter__("group_constrained", someFunc);
cloneState.entities.teams.groupsAssociatedToTeam[teamID].__defineSetter__("ids", someFunc);
```

Doesn't seem worth it.

**Comment by mkraft** [comment]
@hahmadia Good call on the `Object.values` instead of `Object.entries` 👍

File: src/selectors/entities/groups.test.js (hunk 1 above)
Comment (fmunshi) [reply src/selectors/entities/groups.test.js]: I meant something like this @mkraft - up to you either way im 0/5 on it 
```
const modifiedState = {
            ...testState,
            entities: {
                ...testState.entities,
                teams: {
                    ...testState.entities.teams,
                    teams: {
                        ...testState.entities.teams.teams,
                        [teamID]: {
                            ...testState.entities.teams.teams[teamID],
                            group_constrained: true,
                        },
                    },
                    groupsAssociatedToTeam: {
                    ...testState.entities.teams.groupsAssociatedToTeam,
                        [teamID]: {
                            ...testState.entities.teams.groupsAssociatedToTeam[teamID],
                            ids: [expectedAssociatedGroupID1],
                        },
                    },
                },
            },
        };
```

**Comment by srkgupta** [comment]
Waiting for the dependent PR to be Dev approved
https://github.com/mattermost/mattermost-server/pull/14619

**Comment by amyblais** [comment]
@mkraft This one as well needs cherry picking

**Comment by mkraft** [comment]
> @mkraft This one as well needs cherry picking

Done now. Thanks Amy.
